source/common/upstream/outlier_detection_impl.cc

```diff
@@ -82,28 +98,111 @@ Http::Code DetectorHostMonitorImpl::resultToHttpCode(Result result) {
+// are not treated differently. All errors are mapped to HTTP codes.
+// Depending on the value of the parameter *code* the function behaves differently:
+// - if the *code* is not defined, mapping uses resultToHttpCode method to do mapping.
+// - if *code* is zero, no mapping takes place: result is dropped and not reported to outlier
+// detector.
+// - if *code* is non-zero, it is taken as HTTP code and reported as such to outlier detector.
+void DetectorHostMonitorImpl::putResultNoLocalExternalSplit(Result result,
+                                                            absl::optional<uint64_t> code) {
+  if (code) {
+    if (0 == code.value()) {
+      // Zero value of the code parameter indicates NOP. No mapping should take place
+      // between result parameter and HTTP code.
+      return;
+    } else {
+      putHttpResponseCode(code.value());
+    }
+  } else {
+    putHttpResponseCode(enumToInt(resultToHttpCode(result)));
+  }
+}
+
+// Method is called by putResult when external and local origin errors
+// are treated separately. Local origin errors have separate counters and
+// separate success rate monitor.
+void DetectorHostMonitorImpl::putResultWithLocalExternalSplit(Result result,
+                                                              absl::optional<uint64_t>) {
+  switch (result) {
+  // SUCCESS is used to report success for connection level. Server may still respond with
+  // error, but connection to server was OK.
+  case Result::LOCAL_ORIGIN_CONNECT_SUCCESS:
+    return localOriginNoFailure();
+  // Connectivity related errors.
+  case Result::LOCAL_ORIGIN_TIMEOUT:
+  case Result::LOCAL_ORIGIN_CONNECT_FAILED:
+    return localOriginFailure();
+  // EXT_ORIGIN_REQUEST_FAILED is used when connection to server was successful, but transaction on
+  // server level failed. Since it it similar to HTTP 5xx, map it to 5xx handler.
+  case Result::EXT_ORIGIN_REQUEST_FAILED:
+    // map it to http code and call http handler.
+    return putHttpResponseCode(enumToInt(Http::Code::ServiceUnavailable));
+    break;
+  // EXT_ORIGIN_REQUEST_SUCCESS is used to report that transaction with non-http server was
+  // completed successfully. This means that connection and server level transactions were
+  // successful. Map it to http code 200 OK and indicate that there was no errors on connection
+  // level.
+  case Result::EXT_ORIGIN_REQUEST_SUCCESS:
+    putHttpResponseCode(enumToInt(Http::Code::OK));
+    localOriginNoFailure();
+    break;
+  }
+}
+
+// Method is used by other components to reports success or error.
+// It calls putResultWithLocalExternalSplit or put putResultNoLocalExternalSplit via
+// std::function. The setting happens in constructor based on split_external_local_origin_errors
+// config parameter.
+void DetectorHostMonitorImpl::putResult(Result result, absl::optional<uint64_t> code) {
+  put_result_func_(this, result, code);
+}
+
+void DetectorHostMonitorImpl::localOriginFailure() {
```

Comment:
Is it possible to share this code somehow with the other failure function? There is some tricky logic here and I'm wondering if somehow there could be a shared function that is used in both places? Not sure if that is possible or not but just throwing it out there for thought.

Reply:
Not sure what you mean. I went through the code and does not see how they could be shared.